### DISCUSSION TITLE
Git status provider for Azure DevOps is not documented - How do I do it?

### DISCUSSION THREAD

**Discussion by mmhealey1** [fluxcd/flux2#1089] [Q&A]
Looking at the Provider CRD docs ( https://toolkit.fluxcd.io/components/notification/provider/#specification) you can see that is references types in the enum: 
```
// Type of provider
    // +kubebuilder:validation:Enum=slack;discord;msteams;rocket;generic;github;gitlab
    // +required
```

It doesn't mention how to setup azure DevOps but in the guides section it mentions that it is possible - https://toolkit.fluxcd.io/guides/notifications/#git-commit-status

Can someone please tell me what I should put for the type to get this to work. 

Thanks!

**Comment by phillebaba** [comment]
This is just the docs that are no up to date, I will fix this. In the mean time all you need to do is set the type to `azuredevops` and it should work. Hopefully the docs are clear form their on.

**Reply by mmhealey1** [reply]
Thank for this @phillebaba, I am hitting another issue related to this. I am getting the following error when trying to authenticate with my private repo in Azure DevOps - 

{"level":"error","ts":"2021-03-15T11:04:06.457Z","logger":"event-server","msg":"failed to send notification","object":"flux-system/flux-system","kind":"Kustomization","error":"TF400813: The user '' is not authorized to access this resource."}

Now, I have generated a PAT from Azure DevOps and created a base64 encoded secret token in the form <username>:<PAT> as per the docs (the docs mention that it is necessary for BitBucket but this is also true for Azure DevOps). Any ideas what I might be doing wrong? It looks like the username is not being pulled out of then secret.

**Reply by phillebaba** [reply]
Two quick things that has gone wrong for others. First make sure that your PAT has repo:status permissions, otherwise NC will not be allowed to post commit status. The second is to check that your secret does not contain new lines in the base 64 values. For some weird reason that seems to be an issue for a lot of people. Please get back to me if that does not work.

**Reply by mmhealey1** [reply]
@phillebaba They don't appear to be working...

I have checked that the PAT has permission to write status updates using postman and hitting the Azure DevOps api manually (https://dev.azure.com/{organisation}/{project}/_apis/git/repositories/{repo}/commits/{commit}/statuses?api-version=6.0) and that works just fine. I had to set the authentication header to be basic auth and provide the username and password.

I am generating the secret using 
```
echo "$USERNAME:$PAT" | base64  -w 0
```
which removes the newline character that gets generated. And when I view the secret via `kubectl describe secret` it shows the $USERNAME:$PASSWORD as expected. 

I really appreciate you helping me out, I would love to get to the bottom of this :)

**Reply by stealthybox** [reply]
^ additionally  `echo -n "$USERNAME:$PAT"` will completely omit the newline altogether.

**Comment by naustin** [comment]
I installed this connector and started getting reconciliation status updates in my Azure Devops Repo.   https://github.com/Azure/gitops-connector
